ethcore/types/src/transaction/transaction.rs

```diff
@@ -152,59 +212,134 @@ impl HeapSizeOf for Transaction {
+    pub fn encode(&self, signature: Option<&SignatureComponents>) -> Vec<u8> {
+        let mut stream = RlpStream::new();
+        //stream.begin_list(2);
+        //stream.append(&3u8);
+
+        let mut list_size = 7;
+        list_size += if signature.is_some() { 3 } else { 0 };
+        stream.begin_list(list_size);
+        self.transaction.rlp_append_open(&mut stream, None);
+
+        //access list
+        stream.begin_list(self.access_list.len());
+        for access in self.access_list.iter() {
+            stream.begin_list(2);
+            stream.append(&access.0);
+            stream.begin_list(access.1.len());
+            for storage_key in access.1.iter() {
+                stream.append(storage_key);
+            }
+        }
+        if let Some(signature) = signature {
+            signature.rlp_append(&mut stream);
         }
+
+        [&[0x03], stream.as_raw()].concat()
+    }
+
+    pub fn rlp_append(&self, rlp: &mut RlpStream, signature: &SignatureComponents) {
+        rlp.append(&self.encode(Some(signature)));
+    }
+
+    pub fn hash(&self) -> H256 {
+        keccak(&self.encode(None))
     }
 }
 
-impl From<ethjson::transaction::Transaction> for UnverifiedTransaction {
-    fn from(t: ethjson::transaction::Transaction) -> Self {
-        let to: Option<ethjson::hash::Address> = t.to.into();
-        UnverifiedTransaction {
-            unsigned: Transaction {
-                nonce: t.nonce.into(),
-                gas_price: t.gas_price.into(),
-                gas: t.gas_limit.into(),
-                action: match to {
-                    Some(to) => Action::Call(to.into()),
-                    None => Action::Create,
-                },
-                value: t.value.into(),
-                data: t.data.into(),
-            },
-            r: t.r.into(),
-            s: t.s.into(),
-            v: t.v.into(),
-            hash: 0.into(),
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum TypedTransaction {
+    Legacy(Transaction), // old legacy RLP encoded transaction
+    AccessList(AccessListTx), // EIP-2930 Transaction with a list of addresses and storage keys that the transaction plans to access.
+                              // Accesses outside the list are possible, but become more expensive.
+}
+
+//Function that are batched from Transaction struct and needs to be reimplemented
+impl TypedTransaction {
+    pub fn tx_type(&self) -> u8 {
+        match self {
+            Self::Legacy(_) => 0x00,
+            Self::AccessList(_) => 0x03,
         }
-        .compute_hash()
     }
-}
 
-impl Transaction {
     /// The message hash of the transaction.
     pub fn hash(&self, chain_id: Option<u64>) -> H256 {
-        let mut stream = RlpStream::new();
-        self.rlp_append_unsigned_transaction(&mut stream, chain_id);
-        keccak(stream.as_raw())
+        match self {
+            Self::Legacy(tx) => tx.hash(chain_id),
+            Self::AccessList(ocl) => ocl.hash(),
```

Comment:
```suggestion
            Self::AccessList(ocl) => ocl.hash(chain_id),
```

I think that is not explicity specified in 2930, but https://github.com/ethereum/EIPs/blob/master/EIPS/eip-155.md also applies, that is adding the `chain_id` in the RLP only when veryfiying the signature.